### ISSUE TITLE
Highlight all matches when searching.

### ISSUE THREAD

**Issue by jordwalke** [t9md/atom-vim-mode-plus#158]
Emacs has a nice built in search mode where all the matches in the file are highlighted, but the *current* match is highlighted more prominently. This is really nice.

**Comment by t9md** [comment]
Is this a feature request? What's result you wan to get?
In vmp match highlighted is done only visible area of buffer and when scroll event happened it refresh matche highlight accordingly.
Im not sure what feature is lacking in current implementation.

**Comment by jordwalke** [comment]
I'm not seeing all *instances* of the word highlighted - I only see one instance of the search word highlighted. My feature request is to see  all instances of the search pattern highlighted simultaneously in all visible buffers but also to have one special instance highlighted more prominently than all the other instances. That special instance of the search pattern that is highlighted more prominently is exactly the same as what is currently highlighted today in VMP.

**Comment by t9md** [comment]
Are you enabling incremental-search feature?
If you enabled incremental-search, you'll see all instances of search highlighted.
Still not sure your request sorry.
Do you want search highlight persisted after search finished?

**Comment by jordwalke** [comment]
After enter is pressed I would like them to remain highlighted, yes.

**Comment by t9md** [comment]
OK, thats is not emacs spesific feature pure Vim also have hlsearch feature and you can clear it no hlsearch.
I already provided persisted highlight as quick-highlight atom package.
So including part of this package's feature is not so difficult.
I'll try it soon.

**Comment by jordwalke** [comment]
I'll try quick highlight to see if it's what I'm looking for. Hopefully it's compatible with VMP

**Comment by jordwalke** [comment]
Yes, I checked out that package and that looks so cool. It would be awesome to have the highlights persisted like that.

**Comment by t9md** [comment]
After considering to directly implementing to this feature to vmp.
I'm getting feel to low motivation to it because
- To support hlsearch feature, it has to be highlighted across visible buffer(not only for current editor).
- It's need big amount of codes to only highlight search result, and codes are greatly duplicated with my quick-highlight atom package.
- I myself always want to highlight **manually**, since auto-highlighted I  have to clear it manually its tedious work for me.

Do you want to really auto-highlighted, clearing it manually is not tedious work for you?
Or you basically never clear highlight manually and adding highlight manually by quickl-highlight package is more tedious work for you?

I think it's better to support this feature natively in vmp, but want to confirm if its really used by user.
In my stance manually adding highlight for given keyword is better than adding highlight automatically and need to clear it manually.

**Comment by jordwalke** [comment]
I have a mapping so that if I press escape a *second* time (when already in normal mode) the highlight is cleared so that is very natural. I almost never want to clear that search highlight though.

- There is only one active persistent pattern at a time (the current pattern being searched with n/N). If you search for a new pattern, the old persistent highlight is automatically cleared.
- The persistent highlight illustrates where your cursor will jump when you hit n/N.
- If the visual styling is very mild (slightly lighter background color, no border) then it's fairly unobtrusive - I almost never feel the need to clear it.

Since it is some work, you may wish to take a poll from the community here. I have a feeling many people would like this feature. Even Atom's built in find command shows you all of the instances that find is cycling through.

**Comment by jordwalke** [comment]
To be clear I think this is a great feature, but totally understand if you prioritize it lower than other critical fixes. I think of this feature like the awesome tab cycling behavior in VMP - it's not really a built in vim feature you're emulating, it's something VMP does to be *better* than Vim. I understand if you prioritize vim compatibility fixes over this. Either way it's good to have this Issue open to track the feature request.

**Comment by t9md** [comment]
Thanks for detailed explanation.
I think I can implementing it without big mess.
Knowing where jump to next is good reason to support this feature.

**Comment by t9md** [comment]
By the way you use incremental-search feature?
And want to know your main editor is still pure Vim?
Vmp have great original motion like move-to-next-edge. If you haven't tried yet I hope you give it a try!

**Comment by jordwalke** [comment]
I certainly use the incremental search feature (with tab cycling as well). I think just keeping those highlights active would be ideal. Also, when the highlights are persistent, the nature of the experience changes such that the border around them is a little much. The border isn't bad when it's merely incremental-search as it is now, but once they become persistent they are kind of distracting. I might consider unifying incremental-search feature and the persistent search feature (again, only one active persistent search pattern at a time) which might warrant a less dramatic styling around the search matches (slightly darker/lighter background perhaps).

**Comment by jordwalke** [comment]
I've used every Vim emulator ever. Even for really good emulators, I can only last a week before switching back to Vim. With `vim-mode-plus` I've lasted a few weeks now, using it about 80% of the time so it's a sign that something is working here. The key thing that really satisfied that vim *feel* was the work you did on Redo to preserve cursor position. It's one of those things vim converts will appreciate even if they can't articulate it - they'll just unexplainably *feel* that vmp is faithful to Vim.

**Comment by t9md** [comment]
Oh, surprised and happy to hear thanks.
For the styling of hlsearch, I agree with you.
If persisted, more unobtrusive style is better.

**Comment by jordwalke** [comment]
Also, I wonder if flash on `n`/`N` is still needed if all search matches are persistent, and the *current* match is highlighted more prominently.

**Comment by t9md** [comment]
Excerpt from CHANGELOG

- New: highlightSearch to highlight last search pattern across all buffer. #158
  - `highlightSearch` config parameter to enable/disable.
  - `vim-mode-plus:toggle-highlight-search`: to toggle `highlightSearch` config value.
  - `vim-mode-plus:clear-highlight-search`: one-time clearing for highlightSearch. equivalent to `:nohlsearch` in pure Vim.

**Comment by t9md** [comment]
My keymap is this
```
'atom-text-editor.vim-mode-plus.normal-mode':
  'space /': 'vim-mode-plus:toggle-highlight-search'
```

**Comment by t9md** [comment]
Released as v0.21.0.
I liked this feature thanks for suggestion. Hope you also like.
Feedbacks are welcome.

**Comment by jordwalke** [comment]
This is working out pretty well. I noticed a couple of bugs:

1. When I clear via `vim-mode-plus:clear-highlight-search`, and I switch to another tab and back, the search items are no longer cleared.
2. The flash coloring now hides the text. (When flashing the bg color is the same as the fg color). I had to disable flash.

**Comment by t9md** [comment]
For 1.  I also notice and fixed in commit above your comment.
For 2. Thanks for catching this.

First I added z-index to flash color so that flash color come front than highlightSearch. unless this flash color is completely hidden when overlapping highlightSearch color.
But This z-index introduce another issue you reported in 2. flash color hide overlapping text completely when text color is same.
So I made flash color a bit transparent to not hide overlapping text.
